### ISSUE TITLE
Unable to successfully build Ceres-Solver on Windows 10

### ISSUE THREAD

**Issue by rhoffman-gpsw** [ceres-solver/ceres-solver#590]
I have been struggling to build the Ceres-Solver for some time.  In need it to work as a x64 library.  Has anyone successfully built and released a compiled version Using Microsoft Visual Studio version 2017 and/or 2019?

Any help is appreciated!

Thanks,

Robert

**Comment by sandwichmaker** [comment]
@rhoffman-gpsw what is the problem that you are facing? and what version of ceres solver are you using ?

**Comment by rhoffman-gpsw** [comment]
I get glog related missing symbols when I try to link the final library.  Is is for C++ build and it is version 1.14.0 from GitHub.

**Comment by rhoffman-gpsw** [comment]
I'm currently trying to get it to build with Visual Studio version 2017.  It has been awhile (about 6 weeks) since I last tried to get it to build properly.  I spent about a week trying to get it to build before I had to move onto other things.  I am renewing my effort today.

**Comment by sandwichmaker** [comment]
@alexsmac you are the glog expert. Do you know what maybe going on here? also @rhoffman-gpsw it would help if you could share the error log with us.

**Comment by mohankumarSriram** [comment]
@sandwichmaker 
I'm able to build Ceres-Solver in debug mode on Windows 10 using Visual Studio, 2019. I followed the instructions [here](http://ceres-solver.org/installation.html#windows). I'm building from source with the latest source code checked out via git. But I'm facing the issues listed below:

1. The simple_bundle_adjuster binary runs extremely slow.
![slow_bundle_adjustment](https://user-images.githubusercontent.com/22681121/82856753-83fd1480-9f2c-11ea-9269-88ec491f7f93.PNG)
(I have not installed the optional SuiteSparse and CXSparse libraries, not sure if that's the reason for the reduced performance.)

2. The release mode build is not working.
![release_mode_build_problem](https://user-images.githubusercontent.com/22681121/82856809-b0b12c00-9f2c-11ea-828a-a70593a91551.PNG)
I'm not sure if @rhoffman-gpsw is facing the same issue. 

3. Unable to use Ceres in client mode, I'm following the instructions [here](http://ceres-solver.org/installation.html#using-ceres-with-cmake).
Basically, I've set "Ceres_DIR" to the "ceres_bin" folder (which is the out-of-tree build folder generated earlier). But cmake is not able to detect ceres and glog headers specified in helloworld.cc
![client_mode_problem](https://user-images.githubusercontent.com/22681121/82856971-3c2abd00-9f2d-11ea-85ca-b798936545ea.PNG)

Please let me know how we can move forward and fix these issues.

**Comment by sandwichmaker** [comment]
@mohankumarSriram the slow speed is because you are building in debug mode.  I do not know why the windows build is not working, @alexsmac may know what is going on with glog/gflags as well as cmake.

**Comment by sandwichmaker** [comment]
@rhoffman-gpsw are you experiencing the same issues as @mohankumarSriram ?

**Comment by rhoffman-gpsw** [comment]
I was able to resolve my issues.  I am new to using Cmake, and I think that part of setting things up with Cmake wasn't perfect.  I was able to make up for it with my knowledge of using Microsoft Visual Studio 2017.  Also, I was building with GLOG_NO_ABBREVIATED_SEVERITIES, which didn't work for some reason.  I switched to make sure that NOGDI was defined for building libraries and for my target program, then I also had to include the Windows system library: Shlwapi.lib, then I was able to link with no missing symbols for both Debug and for Release.  

Now, I can build and run with no problems.

**Comment by rhoffman-gpsw** [comment]
I will leave this issue open for @mohankumarSriram.
If you need me to close it later, let me know.
Thanks for checking up on me!  Much appreciated!
Also, sorry for taking so long to reply.  I finally got back to this issue today and was able to fix my problems.

**Comment by rhoffman-gpsw** [comment]
My issues were not exactly the same either.  I was using Visual Studio 2017.  @mohankumarSriram  is using VS 2019, and by default uses a more up to date compiler that "might" be causing some of his issues.

**Comment by sandwichmaker** [comment]
Thanks @rhoffman-gpsw. @mohankumarSriram  looks like you are facing 

https://stackoverflow.com/questions/7668200/error-lnk2038-mismatch-detected-for-iterator-debug-level-value-0-doesnt

**Comment by mohankumarSriram** [comment]
Thank you, @sandwichmaker! Your link helped, I got it working. The problem was that both "gflags" and "glog" get built in debug mode by default. I had to build them in release mode and then link their release directories with ceres.

You were right, the release build is considerably faster.
![release_mode_works](https://user-images.githubusercontent.com/22681121/83095991-2a2e5300-a0c2-11ea-9f82-eb0b9a4f2266.PNG)

All tests are passing and I also got Ceres to work in client mode!

We might have to update the installation [doc](http://ceres-solver.org/installation.html#windows) to make it easier for others to install Ceres on Windows. (With Visual Studio, 2019 there is no need to create or generate windows specific projects or solutions - cmake projects are [supported](https://docs.microsoft.com/en-us/cpp/build/get-started-linux-cmake?view=vs-2019) right out of the box.) Please let me know if you would like a PR which adds the alternative mode of installation to the docs.

**Comment by sandwichmaker** [comment]
@mohankumarSriram I am glad to hear things worked out. We would be delighted to have a PR. Please follow the instructions at: http://ceres-solver.org/contributing.html

I am going to close this issue. Looking forward to the PR.

**Comment by PhuDoTuong** [comment]
Hello, I followed instructions at [](https://www.cnblogs.com/Pyrokine/p/14636228.html). I used glog 0.5.0 and VS2017 instead.

**Comment by PhuDoTuong** [comment]
Hello, I followed instructions at https://www.cnblogs.com/Pyrokine/p/14636228.html. Also, I used glog 0.5.0 and **VS2017** instead.

**Comment by melhashash** [comment]
You can use vcpkg to install ceres.
